pkg/controller/tas/topology_ungater.go

```diff
@@ -0,0 +1,300 @@
+
+import (
+	"context"
+	"time"
+
+	"github.com/go-logr/logr"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/util/workqueue"
+	"k8s.io/klog/v2"
+	"k8s.io/utils/ptr"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/event"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+
+	configapi "sigs.k8s.io/kueue/apis/config/v1beta1"
+	kueuealpha "sigs.k8s.io/kueue/apis/kueue/v1alpha1"
+	kueue "sigs.k8s.io/kueue/apis/kueue/v1beta1"
+	"sigs.k8s.io/kueue/pkg/controller/core"
+	utilclient "sigs.k8s.io/kueue/pkg/util/client"
+	"sigs.k8s.io/kueue/pkg/util/parallelize"
+	utilpod "sigs.k8s.io/kueue/pkg/util/pod"
+	utiltas "sigs.k8s.io/kueue/pkg/util/tas"
+)
+
+const (
+	ungateBatchPeriod = time.Second
+)
+
+type topologyUngater struct {
+	client client.Client
+}
+
+type podWithUngateInfo struct {
+	pod        *corev1.Pod
+	nodeLabels map[string]string
+}
+
+var _ reconcile.Reconciler = (*topologyUngater)(nil)
+
+// +kubebuilder:rbac:groups="",resources=pods,verbs=get;list;watch;update;patch;delete
+// +kubebuilder:rbac:groups=kueue.x-k8s.io,resources=workloads,verbs=get;list;watch
+// +kubebuilder:rbac:groups=kueue.x-k8s.io,resources=workloads/status,verbs=get
+
+func newTopologyUngater(c client.Client) *topologyUngater {
+	return &topologyUngater{
+		client: c,
+	}
+}
+
+func (r *topologyUngater) setupWithManager(mgr ctrl.Manager, cfg *configapi.Configuration) (string, error) {
+	podHandler := podHandler{}
+	return TASTopologyUngater, ctrl.NewControllerManagedBy(mgr).
+		Named(TASTopologyUngater).
+		For(&kueue.Workload{}).
+		Watches(&corev1.Pod{}, &podHandler).
+		WithOptions(controller.Options{NeedLeaderElection: ptr.To(false)}).
+		WithEventFilter(r).
+		Complete(core.WithLeadingManager(mgr, r, &kueue.ClusterQueue{}, cfg))
+}
+
+var _ handler.EventHandler = (*podHandler)(nil)
+
+type podHandler struct {
+}
+
+func (h *podHandler) Create(_ context.Context, e event.CreateEvent, q workqueue.TypedRateLimitingInterface[reconcile.Request]) {
+	pod, isPod := e.Object.(*corev1.Pod)
+	if !isPod {
+		return
+	}
+	h.queueReconcileForPod(pod, q)
+}
+
+func (h *podHandler) Update(ctx context.Context, e event.UpdateEvent, q workqueue.TypedRateLimitingInterface[reconcile.Request]) {
+	oldPod, isOldPod := e.ObjectOld.(*corev1.Pod)
+	newPod, isNewPod := e.ObjectNew.(*corev1.Pod)
+	if !isOldPod || !isNewPod {
+		return
+	}
+	h.queueReconcileForPod(oldPod, q)
+	h.queueReconcileForPod(newPod, q)
+}
+
+func (h *podHandler) Delete(_ context.Context, e event.DeleteEvent, q workqueue.TypedRateLimitingInterface[reconcile.Request]) {
+	pod, isPod := e.Object.(*corev1.Pod)
+	if !isPod {
+		return
+	}
+	h.queueReconcileForPod(pod, q)
+}
+
+func (h *podHandler) queueReconcileForPod(pod *corev1.Pod, q workqueue.TypedRateLimitingInterface[reconcile.Request]) {
+	if pod == nil {
```

Comment:
I have refactored this a bit:
- passing object instead of pod to the function to commonize between event handlers
- I don't need to check the scheduling gate (it was just a workaround to check "is it a TAS pod")
- proposed a new Label to use here: https://github.com/kubernetes-sigs/kueue/pull/3271

Reply:
As I mentioned in https://github.com/kubernetes-sigs/kueue/pull/3271#issuecomment-2426461284, we will go with the TAS label solution during the Alpha phase.